olp-cpp-sdk-core/tests/thread/TaskContinuationTest.cpp

```diff
@@ -24,210 +24,396 @@
+
+  continuation.Run();
+
+  ASSERT_EQ(future.wait_for(kMaxWaitMs), std::future_status::ready);
+  const auto result = future.get();
+
+  EXPECT_TRUE(result);
+  ASSERT_EQ(result.GetResult(), counter);
+}
+
+TEST_F(TaskContinuationFixture, CallExecuteAsync) {
+  std::promise<ResponseType<olp::client::HttpResponse>> promise;
+  auto future = promise.get_future();
+
+  auto continuation =
+      olp::thread::TaskContinuation(std::move(scheduler))
+          .Then(
+              [](olp::thread::ExecutionContext, std::function<void(int)> next) {
+                std::thread([=]() { next(1); }).detach();
+              })
+          .Then([](olp::thread::ExecutionContext context, int,
+                   std::function<void(olp::client::HttpResponse)> next) {
+            context.ExecuteOrCancelled([next]() {
+              std::thread([=]() {
+                next(olp::client::HttpResponse(
+                    olp::http::HttpStatusCode::CREATED,
+                    olp::http::HttpErrorToString(
+                        olp::http::HttpStatusCode::CREATED)));
+              }).detach();
+              return olp::client::CancellationToken();
+            });
+          })
+          .Finally([&](ResponseType<olp::client::HttpResponse> response) {
+            promise.set_value(std::move(response));
           });
 
-  cont.Run();
-  std::this_thread::sleep_for(
-      std::chrono::milliseconds(50));  // simulate some delay
+  continuation.Run();
+
+  ASSERT_EQ(future.wait_for(kMaxWaitMs), std::future_status::ready);
+  const auto result = future.get();
+
+  EXPECT_TRUE(result);
+  ASSERT_EQ(result.GetResult().GetStatus(), olp::http::HttpStatusCode::CREATED);
+}
+
+TEST_F(TaskContinuationFixture, CheckNoDeadlockSingleThread) {
+  std::promise<ResponseType<int>> promise;
+  auto future = promise.get_future();
+
+  auto continuation = olp::thread::TaskContinuation(scheduler)
+                          .Then([](olp::thread::ExecutionContext,
+                                   std::function<void(int)> next) { next(1); })
+                          .Then([this](olp::thread::ExecutionContext, int value,
+                                       std::function<void(int)> next) {
+                            ASSERT_EQ(value, 1);
+                            scheduler->ScheduleTask([=]() { next(2); });
+                          })
+                          .Then([this](olp::thread::ExecutionContext, int value,
+                                       std::function<void(int)> next) {
+                            ASSERT_EQ(value, 2);
+                            scheduler->ScheduleTask([=]() { next(3); });
+                          })
+                          .Then([](olp::thread::ExecutionContext, int,
+                                   std::function<void(int)> next) { next(3); })
+                          .Finally([&](ResponseType<int> response) {
+                            promise.set_value(std::move(response));
+                          });
```

Comment:
Didn't you mean `next(4)` in the last `Then()`?
If so, a counter would fit perfectly here